### PR TITLE
New sparse HyperLogLog format

### PR DESCRIPTION
**Pull request by martint**
The old format suffered from some pathological errors that compounded
when merging multiple hyperloglog instances. This new implementation
sacrifices space but retains all information needed to convert to a
dense layout without introducing additional error.

Additionally, we get
- Improved estimation while in the sparse regime due to using 2^26 buckets
  for linear counting
- Support for up to 64k buckets (currently limited by the dense implementation)

Fixes https://github.com/airlift/airlift/issues/261

### CODE REVIEW

File: stats/src/main/java/io/airlift/stats/cardinality/SparseHll.java [hunk 1]
```diff
@@ -86,112 +78,108 @@ public SparseHll(Slice serialized)
...
+    private int encode(long hash)
     {
-        shortHashes = mergeShortHashes(other);
-        numberOfHashes = (short) shortHashes.length;
+        return encode(computeIndex(hash, EXTENDED_PREFIX_BITS), numberOfLeadingZeros(hash, EXTENDED_PREFIX_BITS));
+    }
 
-        overflows = mergeOverflows(other);
-        numberOfOverflows = (short) overflows.length;
+    private int encode(int bucketIndex, int value)
```
Comment (erichwang) [inline stats/src/main/java/io/airlift/stats/cardinality/SparseHll.java]: static

File: stats/src/main/java/io/airlift/stats/cardinality/SparseHll.java [hunk 2]
```diff
@@ -263,175 +223,56 @@ else if (bucketIndex < middleBucketIndex) {
...
+                result[index++] = entries[leftIndex++];
             }
             else if (left > right) {
-                result[index++] = other.shortHashes[rightIndex++];
+                result[index++] = other.entries[rightIndex++];
             }
             else {
-                // values are equal, so pick one arbitrarily
-                result[index++] = shortHashes[leftIndex];
+                int value = Math.max(decodeBucketValue(entries[leftIndex]), decodeBucketValue(other.entries[rightIndex]));
```
Comment (erichwang) [inline stats/src/main/java/io/airlift/stats/cardinality/SparseHll.java]: Nice, we needed that

File: stats/src/main/java/io/airlift/stats/cardinality/SparseHll.java [hunk 3]
```diff
@@ -86,112 +78,108 @@ public SparseHll(Slice serialized)
...
 
-        overflows = mergeOverflows(other);
-        numberOfOverflows = (short) overflows.length;
+    private int encode(int bucketIndex, int value)
+    {
+        return (bucketIndex << VALUE_BITS) | value;
     }
 
-    public DenseHll toDense()
+    private static int decodeBucketIndex(int entry)
```
Comment (erichwang) [inline stats/src/main/java/io/airlift/stats/cardinality/SparseHll.java]: this can be implemented with: decodeBucketIndex(EXTENDED_PREFIX_BITS, entry)?

**Comment by erichwang** [comment]
otherwise looks good

File: stats/src/main/java/io/airlift/stats/cardinality/SparseHll.java (hunk 3 above)
Comment (martint) [inline stats/src/main/java/io/airlift/stats/cardinality/SparseHll.java]: yes, good catch

**Pull request closed without merging**